fix _unsafe_view schema to work with functionalization

bdhirsh opened the pull request:
`_unsafe_view` doesn't have any alias annotations, which seems wrong. It's meant to be a "non-view" to autograd, but it should be a view to everything else (functionalization, and all backends). Since the autograd codegen already goes off a hardcoded list of view ops to figure out what to generate `ADInplaceOrView` kernels for, I don't think this change should break autograd. Will confirm with CI.

This PR should get `_unsafe_view` working for functionalization (needed by LTC tests, and also one of the models traced by executorch. cc @ZolotukhinM)

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #75527
* #76358
* __->__ #79148
* #78820

bdhirsh committed:
fix _unsafe_view schema to work with functionalization

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37014354).

bdhirsh committed:
Update on "fix _unsafe_view schema to work with functionalization"


Differential Revision: [D37014354](https://our.internmc.facebook.com/intern/diff/D37014354)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37014354).

bdhirsh committed:
Update on "fix _unsafe_view schema to work with functionalization"


Differential Revision: [D37014354](https://our.internmc.facebook.com/intern/diff/D37014354)

[ghstack-poisoned]

albanD commented on aten/src/ATen/FunctionalInverses.cpp:
```diff
@@ -228,6 +228,14 @@ Tensor FunctionalInverses::transpose_copy_int_inverse(const Tensor& base, const
     }
 }
 
+Tensor FunctionalInverses::_unsafe_view_copy_inverse(const at::Tensor & base, const at::Tensor & mutated_view, bool reapply_views, at::IntArrayRef size) {
```
The size argument is ignored?

bdhirsh replied:
yep - that's actually true for a bunch of other inverse functions too (like `view_copy_inverse()`). The invariant is that `mutated_view` is already of the same size as the `size` argument, and we want to go in the reverse direction, and make it the same size as the base.

albanD reviewed (approved):
SGTM if CI is happy
We should also be careful perf-wise that this is not regressing anything.

bdhirsh commented:
cc @albanD is there anything specific I should check for w.r.t perf? Since the kernels didn't change, I guess the only thing that could really change is any codegen output that resulted from the schema having aliases. I confirmed that that the codegen output for `_unsafe_view` is the same for the autograd kernel, and also that nothing is generated for the `ADInpaceOrView` kernel.

albanD commented:
Yes generated codegen was the main perf concern here. If that didn't change, that sounds good to me.

bdhirsh committed:
Update on "fix _unsafe_view schema to work with functionalization"

`_unsafe_view` doesn't have any alias annotations, which seems wrong. It's meant to be a "non-view" to autograd, but it should be a view to everything else (functionalization, and all backends). Since the autograd codegen already goes off a hardcoded list of view ops to figure out what to generate `ADInplaceOrView` kernels for, I don't think this change should break autograd. Will confirm with CI.

This PR should get `_unsafe_view` working for functionalization (needed by LTC tests, and also one of the models traced by executorch. cc ZolotukhinM)


Differential Revision: [D37014354](https://our.internmc.facebook.com/intern/diff/D37014354)

[ghstack-poisoned]

bdhirsh committed:
Update on "fix _unsafe_view schema to work with functionalization"

`_unsafe_view` doesn't have any alias annotations, which seems wrong. It's meant to be a "non-view" to autograd, but it should be a view to everything else (functionalization, and all backends). Since the autograd codegen already goes off a hardcoded list of view ops to figure out what to generate `ADInplaceOrView` kernels for, I don't think this change should break autograd. Will confirm with CI.

This PR should get `_unsafe_view` working for functionalization (needed by LTC tests, and also one of the models traced by executorch. cc ZolotukhinM)


Differential Revision: [D37014354](https://our.internmc.facebook.com/intern/diff/D37014354)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37014354).

bdhirsh commented on aten/src/ATen/BatchingRegistrations.cpp:
```diff
@@ -1122,6 +1122,9 @@ TORCH_LIBRARY_IMPL(aten, Batched, m) {
   m.impl("view", view_batching_rule);
+  // From the perspective of vmap, view and _unsafe_view are the same.
+  // (they are only different w.r.t. autograd).
+  m.impl("_unsafe_view", view_batching_rule);
```
cc @zou3519 I had to add a batching rule for `_unsafe_view` to get `test_ops_gradients.py` to pass (before it would use the fallback, but it can't anymore now that `_unsafe_view` is officially a view op)

I also ran `pytest test/test_ops.py test/test_vmap.py -k "reshape"` without error in the functorch repo. But I'll probably need to patch up functorch in any case to add the new batching rule, right?

zou3519 replied:
You will need to add a new batching rule in functorch. Also, vmap may run before autograd (this is the case for the vmap in core), so the batching rule should not turn _unsafe_view into a regular view, right?

bdhirsh committed:
Update on "fix _unsafe_view schema to work with functionalization"

`_unsafe_view` doesn't have any alias annotations, which seems wrong. It's meant to be a "non-view" to autograd, but it should be a view to everything else (functionalization, and all backends). Since the autograd codegen already goes off a hardcoded list of view ops to figure out what to generate `ADInplaceOrView` kernels for, I don't think this change should break autograd. Will confirm with CI.

This PR should get `_unsafe_view` working for functionalization (needed by LTC tests, and also one of the models traced by executorch. cc ZolotukhinM)


Differential Revision: [D37014354](https://our.internmc.facebook.com/intern/diff/D37014354)

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

bdhirsh committed:
Update on "fix _unsafe_view schema to work with functionalization"

`_unsafe_view` doesn't have any alias annotations, which seems wrong. It's meant to be a "non-view" to autograd, but it should be a view to everything else (functionalization, and all backends). Since the autograd codegen already goes off a hardcoded list of view ops to figure out what to generate `ADInplaceOrView` kernels for, I don't think this change should break autograd. Will confirm with CI.

This PR should get `_unsafe_view` working for functionalization (needed by LTC tests, and also one of the models traced by executorch. cc ZolotukhinM)


Differential Revision: [D37014354](https://our.internmc.facebook.com/intern/diff/D37014354)

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge -g

bdhirsh commented:
@pytorchbot merge

janeyx99 commented:
@pytorchbot revert -m "Broke 11.3 tests on trunk and on PR, see https://hud.pytorch.org/pytorch/pytorch/commit/46234df5f12e62b891be4ef4574bfa5380c0ad21" -c ignoredsignal
